base/expr.jl

```diff
@@ -535,71 +535,131 @@ non-termination if the method calls some other method that does not terminate.
+requires to be guaranteed to constant fold a call at compile time. It is
+currently equivalent to the following `setting`s:
 
-This `setting` combines the following other assertions:
 - `:consistent`
 - `:effect_free`
-- `:nothrow`
 - `:terminates_globally`
-and is a convenient shortcut.
+
+!!! note
+    This list in particular does not include `:nothrow`. The compiler will still
+    attempt constant propagation and note any thrown error at compile time. Note
+    however, that by the `:consistent`-cy requirements, any such annotated call
+    must consistently throw given the same argument values.
 
 ---
-# `:total_may_throw`
+# `:total`
 
-This `setting` combines the following other assertions:
+This `setting` is the maximum possible set of effects. It currently implies
+the following other `setting`s:
 - `:consistent`
 - `:effect_free`
+- `:nothrow`
 - `:terminates_globally`
-and is a convenient shortcut.
+- `:notaskstate`
 
-!!! note
-    This setting is particularly useful since it allows the compiler to evaluate a call of
-    the applied method when all the call arguments are fully known to be constant, no matter
-    if the call results in an error or not.
-
-    `@assume_effects :total_may_throw` is similar to [`@pure`](@ref) with the primary
-    distinction that the `:consistent`-cy requirement applies world-age wise rather
-    than globally as described above. However, in particular, a method annotated
-    `@pure` should always be `:total` or `:total_may_throw`.
-    Another advantage is that effects introduced by `@assume_effects` are propagated to
-    callers interprocedurally while a purity defined by `@pure` is not.
+!!! warning
+    `:total` is a very strong assertion and will likely gain additional semantics
+    in future versions of julia (e.g. if additional effects are added and included
+    in the definition of `:total`). As a result, it should be used with care.
+    Whenever possible, prefer to use the minimum possible set of specific effect
+    assertions required for a particular application. In cases where a large
+    number of effect overrides apply to a set of functions, a custom macro is
+    recommended over the use of `:total`.
+
+## Negated effects
+
+Effect names may be prefixed by `!` to indicate that the effect should be removed
+from an earlier meta effect. For example, `:pure !:nothrow` indicates that while
+the call is generally pure, it may however throw.
+
+## Comparison to @pure
+`@assume_effects :consistent :effect_free :terminates_globally` is similar to [`@pure`](@ref) with the primary
+distinction that the `:consistent`-cy requirement applies world-age wise rather
+than globally as described above. However, in particular, a method annotated
+`@pure` should always be at least `:consistent :effect_free :terminates_globally`.
+Another advantage is that effects introduced by `@assume_effects` are propagated to
+callers interprocedurally while a purity defined by `@pure` is not.
```

Comment:
```suggestion
## Comparison to `@pure`
`@assume_effects :foldable` is similar to [`@pure`](@ref) with the primary
distinction that the `:consistent`-cy requirement applies world-age wise rather
than globally as described above. However, in particular, a method annotated
`@pure` should always be at least `:consistent :effect_free :terminates_globally`.
Another advantage is that effects introduced by `@assume_effects` are propagated to
callers interprocedurally while a purity defined by `@pure` is not.
```

Reply:
Thanks. I wrote this before I decided to introduce `foldable` and then I forgot to go back.